actgen/wrappers/atari.py

```diff
@@ -0,0 +1,381 @@
+        return np.rollaxis(np.asarray(observation), 2)
+
+
+class LazyFrames:
+    def __init__(self, frames):
+        """This object ensures that common frames between the observations are only stored once.
+        It exists purely to optimize memory usage which can be huge for DQN's 1M frames replay
+        buffers."""
+        self._frames = frames
+
+    def __array__(self, dtype=None):
+        out = np.concatenate(self._frames, axis=0)
+        if dtype is not None:
+            out = out.astype(dtype)
+        return out
+
+    def __len__(self):
+        return len(self._frames)
+
+    def __getitem__(self, i):
+        return self._frames[i]
+
+
+def wrap_deepmind(
+        env_name='MsPacman',
+        max_episode_steps=None,
+        episode_life=False,
+        clip_rewards=False,
+        frame_skip=4,
+        frame_stack=4,
+        frame_warp=(84, 84),
+        **kwargs):
+    """
+    Wraps a skills Montezuma Env in the wrappers to emulate the deepmind style wrappers seen in:
+    https://github.com/openai/baselines/blob/master/baselines/common/atari_wrappers.py#L275
+    """
+    # Structure looks like:
+    # Wrappers --> SteveEnv -> MonteSkillsEnv => AtariSkillsEnv -> NoopWrapper --> BaseEnv (Gym)
+    # Where the double arrow (=>) is subclassing, the longer arrows (-->) are true wrappers, and
+    # the while the others (->) are attributes (self.env)
+
+    # NOTE: NOOP is done in the internal env in gym_montezuma/montezuma_env.py
+    env = gym.make('{}NoFrameskip-v4'.format(env_name))
+    if noop_wrapper:
+        env = NoopResetEnv(env, noop_max=30)
+    env = WarpFrame(env, frame_warp)
+    env = PyTorchFrame(env)
+
+    if max_episode_steps is not None:
+        env = TimeLimit(env, max_episode_steps=max_episode_steps)
+    if episode_life:
+        env = EpisodicLifeEnv(env)
+
+    if 'FIRE' in env.unwrapped.get_action_meanings():
+        env = FireResetEnv(env)
+
+    if clip_rewards:
+        env = ClipRewardEnv(env)
+
+    env = FrameSkipStackPool(env, skip=frame_skip, stack=frame_stack)
+    return env
+
+
+def make_deepmind_atari(
+        env_name='MsPacman',
+        max_episode_steps=None,
+        episode_life=False,
+        clip_rewards=True,
+        frame_stack=True,
+        scale=False):
+    """Configure environment for DeepMind-style Atari.
+    """
+    env = gym.make('{}NoFrameskip-v4'.format(env_name))
+    env = NoopResetEnv(env, noop_max=30)
+    env = MaxAndSkipEnv(env, skip=4)
+    if max_episode_steps is not None:
+        env = TimeLimit(env, max_episode_steps=max_episode_steps)
+        env = FixedDurationHack(env)
+    if episode_life:
+        env = EpisodicLifeEnv(env)
+    if 'FIRE' in env.unwrapped.get_action_meanings():
+        env = FireResetEnv(env)
+    env = WarpFrame(env)
+    if scale:
+        env = ScaledFloatFrame(env)
+    if clip_rewards:
+        env = ClipRewardEnv(env)
+    if frame_stack:
+        env = FrameStack(env, 4)
```

Comment:
is there somewhere I can read more about why we need all these wrappers?

Reply:
Sure. The place to start would be the [Nature DQN](https://web.stanford.edu/class/psych209/Readings/MnihEtAlHassibis15NatureControlDeepRL.pdf) paper, and after that, I'd recommend the [Revisiting ALE](https://arxiv.org/abs/1709.06009) paper.

Long story short, we want to use similar settings to what other people are commonly using, so our results will tell us something that is relevant to their work.